Mobility screening - New Migration Graph Builder

alexnguyen0512 opened the pull request:
New working builder for mapping from electrodes store with corresponding migration graphs.

alexnguyen0512 committed:
Create migration_graph.py

New migration_graph builder from electrodes

alexnguyen0512 committed:
Update migration_graph.py

acrutt commented on emmet-builders/emmet/builders/mobility_screening/migration_graph.py:
```diff
@@ -0,0 +1,64 @@
+        mg = get_migration_graph(compatibility.process_entries(entries), 
+                                 self.working_ion_entry, 
+                                 ltol=self.ltol, stol=self.stol, angle_tol=self.angle_tol)
+        new_item["mg"] = mg
```
Hi Alex! Can we rename this field from "mg" to "migration_graph"? I tend to try minimize the number of abbreviations used so things make more sense to those unfamiliar with the code.

acrutt commented on emmet-builders/emmet/builders/mobility_screening/migration_graph.py:
```diff
@@ -0,0 +1,64 @@
+class MigrationGraphBuilder(MapBuilder):
+    def __init__(
+        self,
+        materials: MongoStore,
```
This can be removed because the materials store isn't used.

acrutt commented on emmet-builders/emmet/builders/mobility_screening/migration_graph.py:
```diff
@@ -0,0 +1,64 @@
+        angle_tol: float = 15,
+        **kwargs
+    ):
+        self.materials = materials
```
This can be removed because the materials store isn't used.

acrutt commented on emmet-builders/emmet/builders/mobility_screening/migration_graph.py:
```diff
@@ -0,0 +1,64 @@
+        new_item["mg"] = mg
+        return new_item
+    
+    def get_migration_graph(cses, wi_entry, max_distance_min=0.1, max_distance_max=10, ltol=0.4, stol=0.6, angle_tol=15) -> dict:
```
Can you add docstrings and specify input types for your functions? We want to be consistent with the rest of emmet. Here is one example for reference: https://github.com/materialsproject/emmet/blob/main/emmet-builders/emmet/builders/materials/electrodes.py#L51

If you are unfamiliar with typing here is an explanation and we can meet next week to go over it if needed! https://docs.python.org/3/library/typing.html

acrutt commented on emmet-builders/emmet/builders/mobility_screening/migration_graph.py:
```diff
@@ -0,0 +1,64 @@
+        stol: float = 0.6,
+        angle_tol: float = 15,
+        **kwargs
+    ):
```
Please add a docstring for the MigrationGraphBuilder

acrutt commented on emmet-builders/emmet/builders/mobility_screening/migration_graph.py:
```diff
@@ -0,0 +1,64 @@
+        self.materials = materials
+        self.electrodes = electrodes
+        self.tasks = tasks
+        self.working_ion_entry = working_ion_entry
```
Per discussion over slack, let's try seeing if we can extract the working ion entry based on the docs from the electrodes store so that this builder is not limited to only analyzing one type of working ion at a time

alexnguyen0512 committed:
Updated builder inputs

Removed materials store and wi_entry as builder inputs. get wi from electrode doc.

acrutt commented on emmet-builders/emmet/builders/mobility_screening/migration_graph.py:
```diff
@@ -0,0 +1,64 @@
+                                 self.working_ion_entry, 
+                                 ltol=self.ltol, stol=self.stol, angle_tol=self.angle_tol)
+        new_item["mg"] = mg
+        return new_item
```
This is something that requires a bit more work, but we should make sure we have all the relevant fields/info in the output MigrationGraph document. This should also be done using a Document Model (see https://github.com/materialsproject/emmet/blob/main/emmet-core/emmet/core/electrode.py#L149 for an example) where we need to create a new one (e.g. MigrationGraphDoc) for the types of documents created by this builder.

acrutt commented on emmet-builders/emmet/builders/mobility_screening/migration_graph.py:
```diff
@@ -0,0 +1,64 @@
+        materials: MongoStore,
+        electrodes: MongoStore,
+        tasks: MongoStore,
+        target: MongoStore,
```
The convention in emmet is to give the target store a specific descriptive label so it is clear what kind of collection is created. Here it may be appropriate to rename this variable to something like mgraph or migration_graph.

alexnguyen0512 committed:
Update migration_graph.py

changed field labels to match convention and added docstrings

alexnguyen0512 commented:
closed to address comments

acrutt commented on emmet-builders/emmet/builders/mobility_screening/migration_graph.py:
```diff
@@ -0,0 +1,66 @@
+        self.stol = stol
+        self.angle_tol = angle_tol
+        self.migration_graphs = migration_graphs
+        self.wi_entries = {}
```
I think getting a dictionary of the relevant working_ion entries is general enough that it might be useful in other contexts. Would it be possible to add this as a part of utils here? https://github.com/materialsproject/emmet/blob/main/emmet-builders/emmet/builders/utils.py

I'm thinking the most general case would be taking a list of working_ion strings (e.g. as specified here: https://github.com/materialsproject/emmet/blob/main/emmet-builders/emmet/builders/materials/electrodes.py#L44) and then returning a dictionary with the lowest energy entry for each working_ion string.

acrutt replied:
And then for this builder, in the init statement it can get the list of all working ions strings (from WORKING_IONS so this is only hard coded in one place) and apply the function to get all the working ion entries. This would also allow the user to check what working_ion entries will be used before running the builder.
